### PR TITLE
Hotbar command palette + switching

### PR DESCRIPTION
**Pull request by jakolehm**
https://user-images.githubusercontent.com/1446224/115190789-3adc2300-a0f1-11eb-9513-6bab2ab626f4.mov

Requires #2551

### CODE REVIEW

**Comment by jakolehm** [comment]
Added clickable index to prev/next switcher (see bottom left):

image

Click opens hotbar switch command palette.

File: src/common/hotbar-store.ts [hunk 1]
```diff
@@ -34,32 +38,78 @@ export class HotbarStore extends BaseStore<HotbarStoreModel> {
...
+    return this.getByid(this.activeHotbarId);
   }
 
   getByName(name: string) {
     return this.hotbars.find((hotbar) => hotbar.name === name);
   }
 
-  add(hotbar: Hotbar) {
-    this.hotbars.push(hotbar);
+  getByid(id: string) {
```
Comment (aleksfront) [inline src/common/hotbar-store.ts]: `getByid`-> `getById`

File: src/renderer/components/hotbar/hotbar-menu.tsx [hunk 2]
```diff
@@ -25,8 +43,33 @@ export class HotbarMenu extends React.Component<Props> {
     return hotbar.items.map((item) => catalogEntityRegistry.items.find((entity) => entity.metadata.uid === item.entity.uid)).filter(Boolean);
   }
 
+  previous() {
```
Comment (aleksfront) [inline src/renderer/components/hotbar/hotbar-menu.tsx:32]: Maybe better to move logic from `previous()` and `next()` methods into store itself. To keep clean `HotbarMenu` from store internals.

**Review by aleksfront — COMMENTED** [review]
While testing hotbar switcher found some issues:
1. Clicking active hotbar number opens a command pallete with few strange `[object Object]` items.
hotbar object
And there's no "initial" or "default" hotbar in the list

2. Probably related to num 1 but `prev` & `next` buttons isn't working right
next prev buttons

3. Add to hotbar option not working! ⚠️ 
adding to hotbar broken

4. `Remove from hotbar` menu item should be disabled for items not listed inside hotbar.

**Comment by jakolehm** [comment]
> While testing hotbar switcher found some issues:

@aleksfront I think most of these issues are because of hotbar-store didn't migrate properly. I'll tweak this PR to have a separate migration step that fixes existing hotbar stores... meanwhile you can just remove the json file before starting the app.

**Comment by Nokel81** [comment]
Shouldn't the "fromStore" function do validation to make sure it is in the correct state?

**Comment by jakolehm** [comment]
> Shouldn't the "fromStore" function do validation to make sure it is in the correct state?

Maybe but not sure if it's good idea to generate uuids on-the-fly if those are missing.. I think it would make more sense to create proper migration that fixes the persistent data here.

File: package.json [hunk 3]
```diff
@@ -2,7 +2,7 @@
   "name": "kontena-lens",
   "productName": "Lens",
   "description": "Lens - The Kubernetes IDE",
-  "version": "5.0.0-alpha.1",
```
Comment (Nokel81) [inline package.json:5]: Why is this changed in this PR?

File: src/common/hotbar-store.ts [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import { action, comparer, observable, toJS } from "mobx";
 import { BaseStore } from "./base-store";
 import migrations from "../migrations/hotbar-store";
+import { v4 as uuid } from "uuid";
```
Comment (Nokel81) [inline src/common/hotbar-store.ts]: nit: I prefer doing `import "uuid" from "uuid";` and then using it as `uuid.v4()` since that means that when looking at the uses I don't have to come back and checker which version we are importing.

File: src/migrations/hotbar-store/5.0.0-alpha.2.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+  run(store) {
+    const hotbars = (store.get("hotbars") || []) as Hotbar[];
+
+    hotbars.forEach((hotbar) => {
+      if (!hotbar.id) {
+        hotbar.id = uuid();
+      }
+    });
+
+    store.set("hotbars", hotbars);
```
Comment (Nokel81) [inline src/migrations/hotbar-store/5.0.0-alpha.2.ts]: nit: I find assigning to the input confusing when reading code.
```suggestion
    const rawHotbars = (store.get("hotbars") || []) as Hotbar[];

    store.set("hotbars", rawHotbars.map(({ id = uuid.v4(), ...rest }) => ({
      id,
      ...rest,
    }));
```

File: src/renderer/components/hotbar/hotbar-add-command.tsx [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import React from "react";
+import { observer } from "mobx-react";
+import { hotbarStore } from "../../../common/hotbar-store";
+import { CommandOverlay } from "../command-palette";
+import { Input, InputValidator } from "../input";
+
+const uniqueWorkspaceName: InputValidator = {
```
Comment (Nokel81) [inline src/renderer/components/hotbar/hotbar-add-command.tsx]: I don't think this should be called ...workspace...

File: src/renderer/components/hotbar/hotbar-menu.tsx [hunk 7]
```diff
@@ -15,8 +22,19 @@ interface Props {
...
+  componentDidMount() {
+    disposeOnUnmount(this, [
+      reaction(() => hotbarStore.activeHotbarId, () => {
+        Notifications.info(`Hotbar "${hotbarStore.getActive().name}" is now active.`, {
+          id: "active-hotbar",
+          timeout: 5_000
+        });
+      })
+    ]);
+  }
```
Comment (Nokel81) [inline src/renderer/components/hotbar/hotbar-menu.tsx]: I too think that this is a bit obnoxious. Would it be possible to do a sliding animation on the hotbar instead?

File: package.json (hunk 3 above)
Comment (jakolehm) [reply package.json:5]: Migration is not executed without bumping version. Is there better way to achieve this?

Comment (Nokel81) [reply package.json:5]: ah I see, makes sense. I was just confused because this wasn't a release PR

File: src/common/hotbar-store.ts (hunk 4 above)
Comment (jakolehm) [reply src/common/hotbar-store.ts]: Changed (although other parts of codebase are using "v4 as uuid").

File: src/migrations/hotbar-store/5.0.0-alpha.2.ts (hunk 5 above)
Comment (jakolehm) [reply src/migrations/hotbar-store/5.0.0-alpha.2.ts]: Refactored a bit, your suggestion was a bit confusing to my brain so the code is now somewhere in the middle.

File: src/renderer/components/hotbar/hotbar-add-command.tsx (hunk 6 above)
Comment (jakolehm) [reply src/renderer/components/hotbar/hotbar-add-command.tsx]: Good catch, fixed.

File: src/renderer/components/hotbar/hotbar-menu.tsx (hunk 7 above)
Comment (jakolehm) [reply src/renderer/components/hotbar/hotbar-menu.tsx]: Not sure how to do that without hacks. Maybe I'll just remove notifications from this PR?

File: src/renderer/components/hotbar/hotbar-menu.tsx (hunk 2 above)
Comment (jakolehm) [reply src/renderer/components/hotbar/hotbar-menu.tsx:32]: Refactored.

File: src/common/hotbar-store.ts (hunk 1 above)
Comment (jakolehm) [reply src/common/hotbar-store.ts]: Good catch, fixed.

File: src/common/hotbar-store.ts (hunk 4 above)
Comment (Nokel81) [reply src/common/hotbar-store.ts]: thanks, I know. For my code I tend to use it.

File: src/migrations/hotbar-store/5.0.0-alpha.2.ts (hunk 5 above)
Comment (Nokel81) [reply src/migrations/hotbar-store/5.0.0-alpha.2.ts]: okay

File: src/renderer/components/hotbar/hotbar-menu.tsx (hunk 7 above)
Comment (Nokel81) [reply src/renderer/components/hotbar/hotbar-menu.tsx]: Sounds good probably a good idea to have it in a separate PR.

I would do it by rendering all of them but explicitly set the bounding box sizes and the `overflow-display: none` and then use JS powered css vars to change the animation placement. 

At least that is how I would start the impl attempt.

File: src/common/hotbar-store.ts [hunk 8]
```diff
@@ -34,32 +38,98 @@ export class HotbarStore extends BaseStore<HotbarStoreModel> {
...
+    hotbar.id = uuid.v4();
+
+    if (!hotbar.items) {
+      hotbar.items = [];
+    }
+
+    this.hotbars.push(hotbar as Hotbar);
+
+    return hotbar as Hotbar;
   }
```
Comment (Nokel81) [inline src/common/hotbar-store.ts]: You don't need the casts if you create the entry after deconstruction.

Also, except for `render*()` functions I think that the return value should be explicitly written out so that function definitions are fixed points. This also should help when changing the implementation of a function because typescript will fail to compile if you don't satisfy the contract.

Plus it seems like you want the `name` to still be required:

```suggestion
interface HotbarCreateOptions extends Hotbar {
  id?: string;
  items?: HotbarItem[];
}

  add(rawHotbar: HotbarCreateOptions): Hotbar {
    const {
      id = uuid.v4(), // initialize to a new ID if not provided
      items = [], // initialize to empty array if not provided
      name,
    } = rawHotbar;
    const hotbar = { id, items, name };

    this.hotbars.push(hotbar);

    return hotbar;
  }
```

File: src/common/hotbar-store.ts [hunk 9]
```diff
@@ -34,32 +38,98 @@ export class HotbarStore extends BaseStore<HotbarStoreModel> {
...
 
   remove(hotbar: Hotbar) {
```
Comment (Nokel81) [inline src/common/hotbar-store.ts:110]: this should be an action

File: src/renderer/components/hotbar/hotbar-add-command.tsx [hunk 10]
```diff
@@ -0,0 +1,48 @@
+import React from "react";
+import { observer } from "mobx-react";
+import { hotbarStore } from "../../../common/hotbar-store";
+import { CommandOverlay } from "../command-palette";
+import { Input, InputValidator } from "../input";
+
+const uniqueHotbarName: InputValidator = {
+  condition: ({ required }) => required,
+  message: () => `Hotbar with this name already exists`,
```
Comment (Nokel81) [inline src/renderer/components/hotbar/hotbar-add-command.tsx]: nit: doesn't need to be ticks.

File: src/renderer/components/hotbar/hotbar-menu.tsx (hunk 2 above)
Comment (Nokel81) [reply src/renderer/components/hotbar/hotbar-menu.tsx:32]: Probably don't need them to be part of this component then.

**Comment by aleksfront** [comment]
Still can't add anything to newly created hotbar. In my case, items can be added to hotbar num 2.
add to hotbar broken

**Comment by jakolehm** [comment]
> Still can't add anything to newly created hotbar. In my case, items can be added to hotbar num 2.

@aleksfront did you remove the hotbar-store json file? It might be corrupted because initially this PR didn't have migration in place...

File: src/common/hotbar-store.ts [hunk 11]
```diff
@@ -86,18 +92,21 @@ export class HotbarStore extends BaseStore<HotbarStoreModel> {
...
 
-    if (!hotbar.items) {
-      hotbar.items = [];
-    }
+    const hotbar = { id, name, items };
 
     this.hotbars.push(hotbar as Hotbar);
 
     return hotbar as Hotbar;
   }
```
Comment (Nokel81) [inline src/common/hotbar-store.ts]: Don't need these casts

**Comment by aleksfront** [comment]
@jakolehm Removed hotbar store file and still can't add anything into any of hotbars. 🐛

**Comment by Nokel81** [comment]
Also I just remembered that in the catalogue, the "Remove item from hotbar" option exists. IMO this is rather confusing (since it would be better to have that on the hotbar item instead). But if we do want it, then it should only be present when the entity actually has a hotbar item.

Further complicating matters it should be possible to have the same entity have multiple hotbar items so would such a button remove all of them?

**Comment by jakolehm** [comment]
> @jakolehm Removed hotbar store file and still can't add anything into any of hotbars. 🐛

@aleksfront you were right, I didn't push all the fixes I had locally 🤦 Should work now.

File: src/common/hotbar-store.ts [hunk 12]
```diff
@@ -34,32 +44,101 @@ export class HotbarStore extends BaseStore<HotbarStoreModel> {
...
+
+    hotbarStore.activeHotbarId = hotbarStore.hotbars[index].id;
+  }
+
+  switchToNext() {
+    let index = hotbarStore.activeHotbarIndex + 1;
+
+    if (index >= hotbarStore.hotbars.length) {
+      index = 0;
+    }
```
Comment (Nokel81) [inline src/common/hotbar-store.ts:135]: nit: 
```suggestion
	const total = hotbarStore.hotbars.length;
	const current = hotbarStore.activeHotbarIndex;
    const next = (current + 1) % total;
```

File: src/common/hotbar-store.ts [hunk 13]
```diff
@@ -34,32 +44,101 @@ export class HotbarStore extends BaseStore<HotbarStoreModel> {
...
+      this.activeHotbarId = this.hotbars[0].id;
+    }
+  }
+
+  switchToPrevious() {
+    let index = hotbarStore.activeHotbarIndex - 1;
+
+    if (index < 0) {
+      index = hotbarStore.hotbars.length - 1;
+    }
```
Comment (Nokel81) [inline src/common/hotbar-store.ts:124]: nit:
```suggestion
	const total = hotbarStore.hotbars.length;
	const current = hotbarStore.activeHotbarIndex
    const prev = (current + total - 1) % total;
```

File: src/migrations/hotbar-store/5.0.0-alpha.0.ts [hunk 14]
```diff
@@ -2,18 +2,22 @@
...
+import { v4 as uuid } from "uuid";
 
 export default migration({
   version: "5.0.0-alpha.0",
   run(store) {
     const hotbars: Hotbar[] = [];
 
     clusterStore.enabledClustersList.forEach((cluster: any) => {
-      const name = cluster.workspace || "default";
+      const name = cluster.workspace;
```
Comment (Nokel81) [inline src/migrations/hotbar-store/5.0.0-alpha.0.ts:13]: We should remove the workspace from the cluster after this migration.

File: src/renderer/components/+catalog/catalog.tsx [hunk 15]
```diff
@@ -57,7 +57,7 @@ export class Catalog extends React.Component {
   }
 
   addToHotbar(item: CatalogEntityItem) {
-    const hotbar = hotbarStore.getByName("default"); // FIXME
+    const hotbar = hotbarStore.getActive();
 
     if (!hotbar) {
       return;
```
Comment (Nokel81) [inline src/renderer/components/+catalog/catalog.tsx:63]: If there is no hotbar we should tell the user to create one. (Or will this function only be displayed to the user if there are hotbars)

File: src/migrations/hotbar-store/5.0.0-alpha.0.ts (hunk 14 above)
Comment (jakolehm) [reply src/migrations/hotbar-store/5.0.0-alpha.0.ts:13]: Let's keep it here for a while.... just because someone might still want to go back to 4.x series after testing 5.x.

File: src/renderer/components/+catalog/catalog.tsx (hunk 15 above)
Comment (jakolehm) [reply src/renderer/components/+catalog/catalog.tsx:63]: Story for another PR... currently Lens will create a hotbar if none exists (and does not allow to remove last one).

File: src/migrations/hotbar-store/5.0.0-alpha.0.ts (hunk 14 above)
Comment (Nokel81) [reply src/migrations/hotbar-store/5.0.0-alpha.0.ts:13]: okay sounds good. Maybe once we release 5.0?

File: src/renderer/components/+catalog/catalog.tsx (hunk 15 above)
Comment (Nokel81) [reply src/renderer/components/+catalog/catalog.tsx:63]: okay sounds good.

**Comment by jakolehm** [comment]
@aleksfront @nevalla PTAL

**Review by nevalla — APPROVED** [review]
LGTM

How user will know what is currently active hotbar, do we show the name anywhere? For example when switching hotbar, the active hotbar could be highlighted or having `(active)`? WDYT?

In future PR we could also add keyboard shortcuts to easily activate next and previous buttons.

**Comment by jakolehm** [comment]
> LGTM
> 
> How user will know what is currently active hotbar, do we show the name anywhere? For example when switching hotbar, the active hotbar could be highlighted or having `(active)`? WDYT?
> 
> In future PR we could also add keyboard shortcuts to easily activate next and previous buttons.

Good points, let's iterate on follow-up PRs 👍

**Comment by aleksfront** [comment]
Do we want to allow adding same cluster to hotbar many times?
same hotbar items

**Comment by jakolehm** [comment]
> Do we want to allow adding same cluster to hotbar many times?
> same hotbar items

I'm not sure, currently we allow. Anyway that decision/change is out-of-scope for this PR 🙂

**Review by aleksfront — APPROVED** [review]
